### PR TITLE
Pubsub last item cache

### PR DESCRIPTION
**Pull request by zofpolkowska**
This PR introduces following changes to `mod_pubsub`:

- `mnesia` cache backend for `pubsub_last_item` extracted to a separate _cache_ backend module
- _cache_ backend can be configured also to use `rdbms`
- the `mod_pubsub_cache_rdbms` is implemented

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5930.1](https://travis-ci.org/esl/MongooseIM/jobs/460710714) / Erlang 19.3 / small_tests / b79bf94d6a49921186b1e20cdf7a708541516867
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5930/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5930/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-28_10.53.24/lib.mongooseim.logs/run.2018-11-28_10.53.25/suite.log.html)

**Comment by mongoose-im** [comment]
[5931.1](https://travis-ci.org/esl/MongooseIM/jobs/460715464) / Erlang 19.3 / small_tests / cfacce03af2b8c5f348d9ef68b999e4628feed07
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5931/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5931/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-28_11.31.45/lib.mongooseim.logs/run.2018-11-28_11.31.46/suite.log.html)

**Comment by mongoose-im** [comment]
[5943.1](https://travis-ci.org/esl/MongooseIM/jobs/461721487) / Erlang 19.3 / small_tests / 4ac8fc5a18f2a63ef21c2df7617fd489cf4956ce
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5943/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5943/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_11.01.03/lib.mongooseim.logs/run.2018-11-30_11.01.04/suite.log.html)

**Comment by mongoose-im** [comment]
[5945.1](https://travis-ci.org/esl/MongooseIM/jobs/461732003) / Erlang 19.3 / small_tests / 07858010d967c42821b6cda78d0602589dbf699e
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5945/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5945/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_11.52.48/lib.mongooseim.logs/run.2018-11-30_11.52.49/suite.log.html)

**Comment by mongoose-im** [comment]
[5946.1](https://travis-ci.org/esl/MongooseIM/jobs/461747790) / Erlang 19.3 / small_tests / 1b6bbd02f4b12873c191356b9f8a366ec308972d
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5946/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5946/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_12.15.53/lib.mongooseim.logs/run.2018-11-30_12.15.54/suite.log.html)

**Comment by mongoose-im** [comment]
[5947.1](https://travis-ci.org/esl/MongooseIM/jobs/461759689) / Erlang 19.3 / small_tests / 42fe19865eea9126621f2b72b9bfed2af9e417e9
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5947/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5947/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_13.03.08/lib.mongooseim.logs/run.2018-11-30_13.03.10/suite.log.html)

**Comment by mongoose-im** [comment]
[5948.1](https://travis-ci.org/esl/MongooseIM/jobs/461786954) / Erlang 19.3 / small_tests / ef68c10f4989d732a130a8782748399f474ab006
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5948/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5948/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-30_13.57.17/lib.mongooseim.logs/run.2018-11-30_13.57.17/suite.log.html)

File: src/pubsub/mod_pubsub.erl [hunk 1]
```diff
@@ -4007,50 +4004,58 @@ get_max_subscriptions_node(Host) ->
...
+        mod_pubsub_cache_backend:delete_last_item(serverhost(Host), Nidx).
 
 -spec get_cached_item(
         Host    :: mod_pubsub:host(),
           Nidx :: mod_pubsub:nodeIdx())
-        -> undefined | mod_pubsub:pubsubItem().
-get_cached_item({_, ServerHost, _}, Nidx) ->
-    get_cached_item(ServerHost, Nidx);
+        -> false | mod_pubsub:pubsubItem().
+get_cached_item({_, Host, _}, Nidx) ->
```
Comment (fenek) [inline src/pubsub/mod_pubsub.erl]: I have a hunch this line was actually correct and applied to PEP.

File: src/pubsub/mod_pubsub_cache_mnesia.erl [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        {pubsub_last_item,
+        Nidx,
+        ItemId,
+        {os:timestamp(),
+            jid:to_lower(jid:to_bare(Publisher))},
+        Payload}
+    ) of
+        ok -> ok
+    catch
+        exit:{aborded, Reason} -> {error, Reason}
```
[the remark this answers is not in the document]
Comment (michalwski) [reply src/pubsub/mod_pubsub_cache_mnesia.erl]: Looks like our tests don't cover this line :)

[the remark this answers is not in the document]
Comment (fenek) [reply src/pubsub/mod_pubsub_cache_mnesia.erl]: I guess it's very difficult to trigger a failure of dirty operation, unless `meck` is used. :)

File: src/pubsub/mod_pubsub.erl [hunk 3]
```diff
@@ -4007,50 +4004,58 @@ get_max_subscriptions_node(Host) ->
...
+    case proplists:get_value(last_item_cache, Opts, false) of
+        false ->
+            ok;
+        Backend ->
+            start_cache_module(Backend)
+    end.
+
+start_cache_module(Backend) ->
+           gen_mod:start_backend_module(mod_pubsub_cache, [{backend, Backend}],
+                [create_table, delete_table, upsert_last_item, delete_last_item, get_last_item]),
```
[the remark this answers is not in the document]
Comment (zofpolkowska) [reply src/pubsub/mod_pubsub.erl]: @michalwski so the list here should be narrowed to `upsert_last_item`, `delete_last_item`, `get_last_item` ?

[the remark this answers is not in the document]
Comment (michalwski) [reply src/pubsub/mod_pubsub.erl]: correct :)

File: src/pubsub/mod_pubsub.erl [hunk 4]
```diff
@@ -297,28 +297,25 @@ init([ServerHost, Opts]) ->
     {_, State} = init_send_loop(ServerHost),
     {ok, State}.
 
-init_backend(Opts) ->
+init_backend(Opts, Host) ->
```
[the remark this answers is not in the document]
Comment (zofpolkowska) [reply src/pubsub/mod_pubsub.erl]: This was a leftover after some previous attempts :)

**Comment by mongoose-im** [comment]
[5956.1](https://travis-ci.org/esl/MongooseIM/jobs/462761690) / Erlang 19.3 / small_tests / 9cfcb1ff018217666b0ec4c6f902ef6691626c89
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5956/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5956/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-03_11.44.19/lib.mongooseim.logs/run.2018-12-03_11.44.20/suite.log.html)

File: src/pubsub/mod_pubsub_cache.erl [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+%% ------------------------ Backend start/stop ------------------------
+
+-callback start() -> ok.
+
+-callback stop() -> ok.
+
+-callback create_table() -> ok | {error, Reason :: term()}.
+
+-callback delete_table() -> ok | {error, Reason :: term()}.
```
[the remark this answers is not in the document]
Comment (zofpolkowska) [reply src/pubsub/mod_pubsub_cache.erl]: Actually, if we don't need to call `delete_table` on `mod_pubsub_cache_mnesia:stop`, should this callback be removed ?

**Comment by mongoose-im** [comment]
[5957.1](https://travis-ci.org/esl/MongooseIM/jobs/462763203) / Erlang 19.3 / small_tests / f8f8551de50ff482a3ad59db1753395d51cbfee8
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5957/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5957/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-03_12.28.54/lib.mongooseim.logs/run.2018-12-03_12.28.55/suite.log.html)

File: src/pubsub/mod_pubsub_cache.erl [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+%% Behaviour callbacks
+%%====================================================================
+
+%% ------------------------ Backend start/stop ------------------------
+
+-callback start() -> ok.
+
+-callback stop() -> ok.
+
+-callback create_table() -> ok | {error, Reason :: term()}.
```
[the remark this answers is not in the document]
Comment (zofpolkowska) [reply src/pubsub/mod_pubsub_cache.erl]: I suppose if there's no need to delete table  and `create_table` will be a part of `start` callback then we don't

**Comment by mongoose-im** [comment]
[5959.1](https://travis-ci.org/esl/MongooseIM/jobs/462783344) / Erlang 19.3 / small_tests / 641465bc8617c3df0c702c94dcd62db6d44c8393
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5959/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5959/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-03_13.24.42/lib.mongooseim.logs/run.2018-12-03_13.24.42/suite.log.html)

File: src/pubsub/mod_pubsub.erl (hunk 1 above)
Comment (zofpolkowska) [reply src/pubsub/mod_pubsub.erl]: Actually, if out Serverhost is `<<"localhost">>` the argument which is passed to this function is `<<"pubsub.localhost">>`

**Comment by mongoose-im** [comment]
[5960.1](https://travis-ci.org/esl/MongooseIM/jobs/462809227) / Erlang 19.3 / small_tests / b4cfc55092bc76284fe43d29097061ec8df25d18
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5960/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5960/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-03_14.09.53/lib.mongooseim.logs/run.2018-12-03_14.09.54/suite.log.html)

**Comment by mongoose-im** [comment]
[5963.1](https://travis-ci.org/esl/MongooseIM/jobs/462849490) / Erlang 19.3 / small_tests / 9c2c96f34bb7b23adfafc8eee0e074db16c95ceb
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5963/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5963/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-03_16.40.09/lib.mongooseim.logs/run.2018-12-03_16.40.10/suite.log.html)

**Comment by mongoose-im** [comment]
[5966.1](https://travis-ci.org/esl/MongooseIM/jobs/463229508) / Erlang 19.3 / small_tests / b4825726866bd7e777e30ff4dded994757907636
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5966/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5966/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-04_09.43.15/lib.mongooseim.logs/run.2018-12-04_09.43.16/suite.log.html)

File: src/pubsub/mod_pubsub_cache_mnesia.erl [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    ok.
+
+-spec upsert_last_item(ServerHost :: binary(),
+                       Nidx :: mod_pubsub:nodeIdx(),
+                       ItemID :: mod_pubsub:itemId(),
+                       Publisher::jid:ljid(),
+                       Payload::mod_pubsub:payload()) -> ok | {error, Reason :: term()}.
+upsert_last_item(_ServerHost, Nidx, ItemId, Publisher, Payload) ->
+    try mnesia:dirty_write(
+        {pubsub_last_item,
```
Comment (fenek) [inline src/pubsub/mod_pubsub_cache_mnesia.erl]: Creating a record with manual tuple construction is evil.

https://github.com/esl/MongooseIM/blob/master/include/pubsub.hrl#L155

**Comment by mongoose-im** [comment]
[5968.1](https://travis-ci.org/esl/MongooseIM/jobs/463295275) / Erlang 19.3 / small_tests / f72a0ae59d91f92421aeaf68eee859214a57bbe3
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5968/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5968/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-04_11.35.23/lib.mongooseim.logs/run.2018-12-04_11.35.23/suite.log.html)

File: src/pubsub/mod_pubsub_cache_mnesia.erl [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    ) of
+        ok -> ok
+    catch
+        exit:{aborted, Reason} -> {error, Reason}
+    end.
+
+-spec get_last_item(ServerHost :: binary(),
+                    Nidx :: mod_pubsub:nodeIdx()) -> [mod_pubsub:pubsubLastItem()] | {error, Reason :: term()}.
+get_last_item(_ServerHost, Nidx) ->
+    try mnesia:dirty_read({pubsub_last_item, Nidx}) of
```
Comment (michalwski) [inline src/pubsub/mod_pubsub_cache_mnesia.erl]: I wonder what if there is no items in the node yet? I think this is a correct situation and in such a case I think it'd be good to return sth like `{error, no_items}`. Do we have a test covering such a scenario?

File: src/pubsub/mod_pubsub_cache_rdbms.erl [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            "created_at = ", EscCreatedAt, ", ",
+            "payload = ", EscPayload,
+            ";"].
+
+
+%%====================================================================
+%% Helpers
+%%====================================================================
+
+check_rdbms_response({selected, [SelectedItem]}) ->
```
Comment (michalwski) [inline src/pubsub/mod_pubsub_cache_rdbms.erl]: Again, what if there was no items yet?

File: src/pubsub/mod_pubsub_cache_rdbms.erl [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    LastItem = item_to_record(SelectedItem),
```
Comment (fenek) [inline src/pubsub/mod_pubsub_cache_rdbms.erl]: The purpose of this function is to "check", not "convert". Either this step should be reflected in function name or done in the function that actually does the query (and would be consistent with the way it is done in main DB backend.

**Comment by mongoose-im** [comment]
[5969.1](https://travis-ci.org/esl/MongooseIM/jobs/463330176) / Erlang 19.3 / small_tests / 9d17397d611464e5340d2009de08bc33cf2b33dc
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5969/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5969/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-04_13.05.18/lib.mongooseim.logs/run.2018-12-04_13.05.18/suite.log.html)

**Comment by mongoose-im** [comment]
[5970.1](https://travis-ci.org/esl/MongooseIM/jobs/463349167) / Erlang 19.3 / small_tests / 086ff9ce680646a6a62f0a5fbddab4b341237238
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5970/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5970/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-04_13.49.27/lib.mongooseim.logs/run.2018-12-04_13.49.28/suite.log.html)

**Comment by mongoose-im** [comment]
[5972.1](https://travis-ci.org/esl/MongooseIM/jobs/463365162) / Erlang 19.3 / small_tests / 65398999dfc0286687f71dea7139b6e34a69bd34
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5972/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5972/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-04_14.43.32/lib.mongooseim.logs/run.2018-12-04_14.43.33/suite.log.html)

**Comment by mongoose-im** [comment]
[5974.1](https://travis-ci.org/esl/MongooseIM/jobs/463385960) / Erlang 19.3 / small_tests / 89c352ccbbbeeb454315bad12cda3c5149354984
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2144/5974/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2144/5974/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-12-04_15.10.38/lib.mongooseim.logs/run.2018-12-04_15.10.39/suite.log.html)

**Pull request merged**